File: julia/count-primes.jl
using Printf 

function countPrimes(n)
  primeNumbers = map(i->true, 1:n-1)
	primeNumbers[1] = false
	last = Int(trunc(sqrt(n)))
	for i in 2:last
		if !primeNumbers[i]
			continue
		end
		j = i * 2
		while j < n 
			primeNumbers[j] = false
			j += i
		end
	end
	c = 0
	for i in primeNumbers
		if i
			c += 1
		end
	end
	return c
end

function main()
	durations = map(i->0.0, 1:100)
	for i in 1:length(durations)
		start = time()
		n = countPrimes(1000000)
		durations[i] = (time() - start) * 1000
		if n != 78498
			@printf("found %d instead of 78498 prime numbers\n", n)
			return nothing
		end
		@printf("Iteration %3d completed in %0.2fms\n", i, durations[i])
	end

	@printf("%d iterations completed in %0.2fs, min: %0.2fms max: %0.2fms average: %0.2fms", 
					length(durations), sum(durations) / 1000, minimum(durations), maximum(durations),
					sum(durations) / length(durations))
end

main()

File: c/count-primes.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int countPrimes(int n) {
	char *primeNumbers = malloc(n);
	int i, j, c;
	for (i = 0; i < n; i++) {
		primeNumbers[i] = 1;
	}
	primeNumbers[0] = 0;
	primeNumbers[1] = 0;
	int last = (int) sqrt((double) n);
	for (i = 2; i < last; i++) {
		if (!primeNumbers[i]) {
			continue;
		}
		for (j = i * 2; j < n; j += i) {
			primeNumbers[j] = 0;
		}
	}
	c = 0;
	for (i = 0; i < n; i++) {
		if (primeNumbers[i]) {
			c++;
		}
	}
	free(primeNumbers);
	return c;
}

int main() {
	int i, n;
	clock_t start, end;
	double min, max, total, avg;
	double *durations = malloc(sizeof(double) * 100);
	
	for (i = 0; i < 100; i++) {	
		start = clock();
		n = countPrimes(1000000);
		end = clock();
		if (n != 78498) {
			fprintf(stderr, "found %d instead of 78498 prime numbers\n", n);
			return -1;
		}
		durations[i] = ((double) (end - start)) / (CLOCKS_PER_SEC / 1000.0);
		printf("Iteration %3d completed in %0.2lfms\n", i, durations[i]);
	}

	max = durations[0];
	min = durations[0];
	total = durations[0];
	for (i = 1; i < 100; i++) {
		if (durations[i] < min) {
			min = durations[i];
		}

		if (durations[i] > max) {
			max = durations[i];
		}

		total += durations[i];
	}
	avg = total / 100.0;
	free(durations);
	printf("%d iterations completed in %0.2lfs, min: %0.2lfms max: %0.2lfms average: %0.2lfms\n",
			100, total / 1000, min, max, avg);
}
